Add Char.bits, {Array,Bytes,String}.filter functions

mauny opened the pull request:
This PR adds `filter` functions to modules `Array`, `Bytes` and `String`, using bit vectors (bytes) to identify elements to be kept in the result.
Char.bits is a helper function that returns the number of bits set in the binary representation of a character.

These `filter` functions follow @damiendoligez's suggestion in the discussion about #882. Benchmarks show that they are more efficient than recursive Array.filter (by @alainfrisch) when the result has more than 20 elts. See https://gist.github.com/mauny/ce78bd7539b3b762296fc4a29097ef7a for a test program and its output (at the end of the file).

mauny committed:
Add Char.bits, {Array,Bytes,String}.filter functions

Add `filter` functions to modules `Array`, `Bytes` and `String`,
using bit vectors to identify elements to be kept in the result.
Char.bits is a helper function that returns the number of bits set in the
binary representation of a character.

xclerc commented on stdlib/char.ml:
```diff
@@ -74,3 +76,40 @@ type t = char
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \003\004\004\005\004\005\005\006\004\005\005\006\005\006\006\007\
+   \003\004\004\005\004\005\005\006\004\005\005\006\005\006\006\007\
+   \004\005\005\006\005\006\006\007\005\006\006\007\006\007\007\b"
```
For the sake of readability, you might want to rewrite the last element from
`\b` to `\008` (I understand you copy/pasted a toplevel output...).

xclerc commented on stdlib/char.ml:
```diff
@@ -74,3 +76,40 @@ type t = char
+
+(* Return the number of bits sets in a char *)
+let bits c =
+ code (string_unsafe_get bitcount (code c))
```
It may be interesting to do some benchmarking in order to assess
whether a table lookup is the best implementation. Hacker's delight
is your friend here; for example [1] has multiple implementations of
population count that are neither loop- nor table-based.

[1] http://www.hackersdelight.org/hdcodetxt/pop.c.txt

bluddy replied:
Alternatively, the table has size 256 bytes, which is 4 cache lines on x86. Since the max value is 8, you could fit it into 2 cache lines if you use nibbles.

alainfrisch replied:
Also https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer

alainfrisch replied:
(Of course, for a byte, this is even simpler.)

xclerc replied:
Indeed, see `pop7` in [1] (2 mult, 2 shift, 1 and).

xclerc commented on stdlib/bytes.ml:
```diff
@@ -314,6 +314,70 @@ let rcontains_from s i c =
     try ignore (rindex_rec s i c); true with Not_found -> false
 
+let filter p arg =
+  let size_arg = length arg in
```
Isn't it possible to share part of the logic with `Array`?
(At least the conversion of eight booleans into a char.)

bluddy replied:
But where would the function go? You'd be adding a dependency between basic library modules.

xclerc replied:
If we cannot (or don't want to) share part of the logic,
we should probably add comments to indicate that any
modification should be replicated in the other module.

alainfrisch commented on stdlib/char.mli:
```diff
@@ -52,6 +52,11 @@ val uppercase_ascii : char -> char
    @since 4.03.0 *)
 
+val bits : char -> int
+(** Return the number of bits set to 1 in the bnary representation of the
```
Typo "bnary".  Also, I'm not sure about the name.  For instance, it has nothing to do with Int64.bits_of_float.  Common names for the operation: Hamming weight, popcount.

damiendoligez replied:
... or with `Random.bits`.

My vote goes to `popcount`.

alainfrisch commented on stdlib/array.ml:
```diff
@@ -292,3 +292,68 @@ let stable_sort cmp a =
+    (* create the bit vector *)
+    let bv =
+      Bytes.create
+        (if size_arg land 7 == 0 then size_arg lsr 3
```
`(size_arg - 1) lsr 3 + 1` is simpler and probably a bit faster.

alainfrisch commented:
I think it is useful to have an efficient implementations of Array.filter in the stdlib.

The approach used here could benefit from introducing a BoolArray (or BitArray) modules, implementing the "Array signature" on type bool.  The first pass would just be a simple call to BoolArray.init.  I'm wondering what would be the performance penalty of using this abstraction.  Obviously, BoolArray would have plenty of other uses. (We already have custom char arrays and soon float arrays, so having a dedicated bool array implementation does not sound crazy.)

dbuenzli commented:
> I'm wondering what would be the performance penalty of using this abstraction

If someone wants to measure that [here's](https://github.com/dbuenzli/uucp/blob/master/src/uucp_tmapbool.ml) a use case. The package has performance tests.

xclerc commented on stdlib/array.ml:
```diff
@@ -292,3 +292,68 @@ let stable_sort cmp a =
+        let c = Char.code (Bytes.unsafe_get bv i_bv) in
+        let base = i_bv lsl 3 in
+        for i = 0 to 7 do
+          if c land (1 lsl i) <> 0 then begin
```
For the record, there is a primitive (namely `Pbittest`, with corresponding
C function `caml_bitvect_test`) that performs this kind of computation, but
your code is probably a tad faster.

mauny committed:
Rename Char.bits into Char.popcount
Fix typos

mauny commented:
Thanks for your comments. Here is what I did

* changed `\b` to `\008`, and corrected typos
* renamed Char.bits to Char.popcount
* changed `(size_arg lsr 3)` into `(size_arg - 1) lsr 3 + 1` when `size_arg land 7 == 0`
* About sharing parts of `Array.filter` and `Bytes.filter`, I followed @xclerc's last suggestion and added a comment in both files in order to indicate that the `filter` functions are essentially the same source code.
* About the implementation of `popcount`, I tried several versions, from the references that @xclerc and @alainfrisch gave, but:
     1. in my tests, they have no visible influence on the efficiency of `Array.filter`
     2. when micro-benchmarking them, the implementation with a table is 7x more efficient than versions with shift/add operations.
I suggest to keep the current version for the time being. Some compiler support (POPCNT instructions or __builtin_popcount provided by GCC, see e.g. https://www.privacore.com/2016/03/17/popcnt-instruction-adventure/) could improve the efficiency of popcount functions in the future, but this should be part of another PR.
* Finally, following @alainfrish's suggestion, I am currently testing a Bitvect/BoolArray module to measure the cost of abstracting bitvect operations. More in a few days.

mauny commented:
Appveyor is failing, and this seems related to windows (run_win32.c(61)).

chambart commented on stdlib/array.ml:
```diff
@@ -292,3 +292,72 @@ let stable_sort cmp a =
+        (if p (unsafe_get arg (i8 lor 4)) then  16 else 0) lor
+        (if p (unsafe_get arg (i8 lor 5)) then  32 else 0) lor
+        (if p (unsafe_get arg (i8 lor 6)) then  64 else 0) lor
+        (if p (unsafe_get arg (i8 lor 7)) then 128 else 0) in
```
Is that sequence better than a for loop ?
```OCaml
let accum = ref 0 in
for i = 0 to 7 do
  if p (unsafe_get arg (i8 lor i)) then accum := !accum lor (1 lsl i)    
done
```
The reduction in code size might be beneficial, and the single call to `p` might make the filter function more amenable to inlining. When the predicate is not inlined, the benefit from unrolling that loop might be lost in the noise.

mauny replied:
On my benchmarks (without flambda and where `p` tests `n mod k`), a `for` loop is 7% slower than my original code (on average).

I understand your point about inlining, but since Array.filter is meant to be a library function, inlining of `p` will happen only if `Array.filter` gets inlined itself, isn't it? Please confirm that it may happen and I'll make the change.

Thanks,

xclerc commented:
Thanks for both the changes and the benchmarks!

It is interesting to know that there is no visible difference
between the various implementations. (Regarding the micro
benchmark, I guess it does not account for the fact in real-life
the table may be evicted from the cache.)

As you point out, the prospect of a primitive-based implementation
(which is why I mentioned `Pbittest`), should be left to another PR.

alainfrisch commented:
> Finally, following @alainfrish's suggestion, I am currently testing a Bitvect/BoolArray module to measure the cost of abstracting bitvect operations. More in a few days.

Any news from this part?

As for naming, note that the preferred direction would now be to create a new Bool module (under Stdlib), with an Array submodule.

xavierleroy commented:
18 months later, any news?

alainfrisch commented:
I think the functions are very useful additions and good fit for the stdlib.  I'm mostly satisfied with the current state, and I'm planning to review it in detail, hopefully soon.

alainfrisch commented:
To be continued in #9049.
